### PR TITLE
Fix logging's step values when multiple dataloaders are used during evaluation

### LINKED ISSUE #11616: Inconsistent step values when using multiple val/test dataloaders when logged on step-level
## 🐛 Bug

Currently, the `LoggerConnector` handles the val_step index for step-level logging for val/test steps, due to which it shows incorrect step (x-axis) in case of multiple dataloaders.

### To Reproduce

<details><summary>CODE</summary>
<p>

```py
import torch
from torch.utils.data import DataLoader, Dataset
import time

from pytorch_lightning import LightningDataModule, LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size: int, length: int):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        return {"loss": loss}

    def validation_step(self, batch, batch_idx, dataloader_idx=None):
        loss = self(batch).sum()
        self.log('valid_loss', loss, on_step=True)

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        return optimizer

    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 64))

    def val_dataloader(self):
        return [DataLoader(RandomDataset(32, 64)), DataLoader(RandomDataset(32, 64))]
```
</p>
</details>

```py
model = BoringModel()
trainer = Trainer(max_epochs=1)
trainer.fit(model)
```

Output:
The log for dataloader_idx=1 starts from 64, which is the step right after the first dataloader is completed.

Screenshot 2022-01-25 at 5 10 43 PM

Expected:
for each datalaoder, it should start from 0.

If we go a little more crazy and do:

```py
model = BoringModel()
trainer = Trainer(max_epochs=1)
trainer.fit(model)
trainer.validate(model)
```

Output:
Screenshot 2022-01-25 at 7 17 09 PM

Expected:
I think it should replace the logs created during fit in such a case.

### Proposal
If we consider the above issues:
I think a better solution would be to let `EvaluationLoop` track the logging step for each dataloader, instead of logger connector, in a list where each index represents the current dataloader and will be the only one that should be incremented.

But there can be another issue, where the user resets the dataloader after every n epochs, then in such a case, should we reset the list if we go with this? The reason is, we can't for sure say whether the new dataloaders will be the same/more/less, and if we don't reset, it might be a problem.

### Environment

- PyTorch Lightning Version (e.g., 1.5.0): master
- PyTorch Version (e.g., 1.10): 1.10
- Python version (e.g., 3.9): 3.0
- OS (e.g., Linux): mac
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### Additional context

Open to better alternative solutions.

cc @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @justusschock @ninginthecloud

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #11616

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @otaj @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @justusschock @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 1]
```diff
@@ -137,35 +134,15 @@ def log_metrics(self, metrics: _OUT_DICT, step: Optional[int] = None) -> None:
...
     def _evaluation_epoch_end(self) -> None:
         results = self.trainer._results
         assert results is not None
         results.dataloader_idx = None
 
-    def update_eval_step_metrics(self) -> None:
+    def update_eval_step_metrics(self, step: int) -> None:
         assert not self._epoch_end_reached
-        if self.trainer.sanity_checking:
-            return
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:162]: Why remove this?

File: pytorch_lightning/loops/fit_loop.py [hunk 2]
```diff
@@ -218,7 +218,8 @@ def reset(self) -> None:
     def on_run_start(self) -> None:  # type: ignore[override]
         """Calls the ``on_train_start`` hook."""
         # reset train dataloader and val dataloader
-        self.trainer.reset_train_val_dataloaders(self.trainer.lightning_module)
+        self.trainer.reset_train_dataloader(self.trainer.lightning_module)
+        self.epoch_loop.val_loop._reload_evaluation_dataloaders()
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:222]: Can we defer this call to the moment validation loop actually runs?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:162]: moved to loop instead where we are managing the step values. Since we don't need to increment the step during sanity check

File: pytorch_lightning/loops/fit_loop.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/fit_loop.py:222]: sorry, what do you mean? I made this change for the case when `num_sanity_steps=0`, the validation dataloaders are loaded here, but we need to use the loop method to load them to ensure that the tracker is updated correctly.

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 3]
```diff
@@ -924,3 +925,57 @@ def test_rich_print_results(inputs, expected):
...
+    trainer.test(model)
+
+    for dl_idx in range(num_dataloaders):
+        suffix = f"/dataloader_idx_{dl_idx}" if num_dataloaders == 2 else ""
+        assert logger.logged_metrics[f"val_log_fit{suffix}"]["logged_values"] == list(range(limit_batches)) * max_epochs
+        assert logger.logged_metrics[f"val_log_fit{suffix}"]["logged_steps"] == list(range(limit_batches * max_epochs))
+        assert logger.logged_metrics[f"val_log_validate{suffix}"]["logged_values"] == list(range(limit_batches))
+        assert logger.logged_metrics[f"val_log_validate{suffix}"]["logged_values"] == list(range(limit_batches))
+        assert logger.logged_metrics[f"test_log{suffix}"]["logged_steps"] == list(range(limit_batches))
+        assert logger.logged_metrics[f"test_log{suffix}"]["logged_steps"] == list(range(limit_batches))
```
Comment (ethanwharris) [inline tests/trainer/logging_/test_eval_loop_logging.py:981]: Some duplication here:
```suggestion
        assert logger.logged_metrics[f"val_log_validate{suffix}"]["logged_values"] == list(range(limit_batches))
        assert logger.logged_metrics[f"test_log{suffix}"]["logged_steps"] == list(range(limit_batches))
```

**Review by ethanwharris — APPROVED** [review]

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 3 above)
Comment (rohitgr7) [reply tests/trainer/logging_/test_eval_loop_logging.py]: good catch!!

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 4]
```diff
@@ -219,10 +220,17 @@ def _get_max_batches(self) -> List[int]:
...
+        dataloaders = None
+
         if self.trainer.testing:
             self.trainer.reset_test_dataloader()
+            dataloaders = self.trainer.test_dataloaders
         elif self.trainer.val_dataloaders is None or self.trainer._data_connector._should_reload_val_dl:
             self.trainer.reset_val_dataloader()
+            dataloaders = self.trainer.val_dataloaders
+
+        if dataloaders:
```
Comment (justusschock) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:232]:
```suggestion
        if dataloaders is not None:
```

just a nit: but let's be precise :)

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 4 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:232]: did this way to make sure `[]` are handled too

File: pytorch_lightning/loops/fit_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/loops/fit_loop.py:209]: My question is, why call this here in `FitLoop.on_run_start` instead of on `FitLoop.EpochLoop.EvaluationLoop.on_run_start`?

File: pytorch_lightning/loops/fit_loop.py [hunk 5]
```diff
@@ -205,7 +205,8 @@ def reset(self) -> None:
     def on_run_start(self) -> None:  # type: ignore[override]
         """Calls the ``on_train_start`` hook."""
         # reset train dataloader and val dataloader
-        self.trainer.reset_train_val_dataloaders(self.trainer.lightning_module)
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:208]: This method is now unused. Deprecate it?

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 6]
```diff
@@ -929,3 +930,57 @@ def test_rich_print_results(inputs, expected):
...
+def test_eval_step_logging(tmpdir, num_dataloaders):
+    class CustomLogger(CSVLogger):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            self.logged_metrics = collections.defaultdict(lambda: collections.defaultdict(list))
+
+        def log_metrics(self, metrics, step):
+            for key in metrics:
+                self.logged_metrics[key]["logged_values"].append(metrics[key])
+                self.logged_metrics[key]["logged_steps"].append(step)
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:945]: Using the CSV logger to test this is a bit ugly.

Why not mock `logger.log_metrics` instead and test the call inputs?

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 7]
```diff
@@ -287,3 +291,10 @@ def _should_track_batch_outputs_for_epoch_end(self) -> bool:
         if self.trainer.testing:
             return is_overridden("test_epoch_end", model)
         return is_overridden("validation_epoch_end", model)
+
+    def _initialize_batch_idx_tracker(self, num_dataloaders: int) -> None:
+        self._seen_batches = [0] * num_dataloaders
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:296]: shouldn't this initialize with the existing ptracking value in case of reload?

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 8]
```diff
@@ -287,3 +291,10 @@ def _should_track_batch_outputs_for_epoch_end(self) -> bool:
...
+
+    def _increment_batch_idx_tracker(self, dataloader_idx: int) -> None:
+        if not self.trainer.sanity_checking:
+            self._seen_batches[dataloader_idx] += self.batch_progress.current.completed
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:300]: Why not do this one by one inside `advance` where you get the `dl_eval_step`?

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 9]
```diff
@@ -135,7 +136,10 @@ def advance(  # type: ignore[override]
         self.batch_progress.increment_completed()
 
         # log batch metrics
-        self.trainer._logger_connector.update_eval_step_metrics()
+        if not self.trainer.sanity_checking:
+            dl_idx = kwargs.get("dataloader_idx", 0)
+            dl_eval_step = self._seen_batches[dl_idx] + self.batch_progress.current.completed - 1
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:141]: Could this just be a += 1?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 10]
```diff
@@ -37,8 +36,6 @@ def __init__(self, trainer: "pl.Trainer", log_gpu_memory: Optional[str] = None)
                 "Please monitor GPU stats with the `DeviceStatsMonitor` callback directly instead."
             )
         self.log_gpu_memory = log_gpu_memory
-        self._val_log_step: int = 0
-        self._test_log_step: int = 0
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:41]: Nice

File: pytorch_lightning/loops/fit_loop.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/fit_loop.py:209]: for the progress bar. The main progress bar requires the total batches for both which is created during training.

File: pytorch_lightning/loops/fit_loop.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/fit_loop.py:208]: follow up PR! Will be quick if this PR get's merged quickly today.

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 7 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:296]: what if the number of dataloaders changes? if they are new dataloaders, I'd suggest reinitializing them

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 6 above)
Comment (rohitgr7) [reply tests/trainer/logging_/test_eval_loop_logging.py:945]: can be done that way too. will update

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 8 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:300]: I thought since we already have a tracker, why not use that? batch_progress resets after each dataloader call.

**Review by carmocca — APPROVED** [review]
I included the deprecation here. It can be easily removed when cherry-picked into a bugfix release.

File: CHANGELOG.md [hunk 11]
```diff
@@ -235,6 +235,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed issue where the CLI could not pass a `Profiler` to the `Trainer` ([#13084](https://github.com/PyTorchLightning/pytorch-lightning/pull/13084))
 
 
+- Fixed logging on step level for eval mode ([#12184](https://github.com/PyTorchLightning/pytorch-lightning/pull/12184))
```
Comment (awaelchli) [inline CHANGELOG.md:238]: This is too general. You need to mention that this is fixed for the case of multiple dataloaders. As I understand, there is no known bug for single dataloader use cases.

**Review by awaelchli — APPROVED** [review]
The fact that lightning offers step level logging for validation/testing still baffles me today. Absolute insanity.
Anyway, thanks @rohitgr7 for working on this!

File: CHANGELOG.md (hunk 11 above)
Comment (carmocca) [reply CHANGELOG.md:238]: What about "Fixed logging's step values when multiple dataloaders are used during evaluation"?

I'll resolve this comment and change it in a separate PR as it's ready to go otherwise. Save the trees :evergreen_tree:
